Make FixedChannelPool#close() wait for all idle channels to be closed

lutovich opened the pull request:
Motivation:

`FixedChannelPool#close()` is a blocking call when executed by non event loop thread. It offloaded closing of the idle channels to `GlobalEventExecutor` and could return before all channels were actually closed.

Modifications:

Introduced an async method to close all idle channels in the parent class `SimpleChannelPool`. `FixedChannelPool.close()` uses it in a blocking fashion when invoked by a non event loop thread.

Result:

`FixedChannelPool.close()` returns to the caller only after all idle channels are closed.

Fixes #8398.

normanmaurer commented on transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java:
```diff
@@ -389,13 +392,43 @@ protected boolean offerChannel(Channel channel) {
 
     @Override
     public void close() {
+        closeIdleChannels().awaitUninterruptibly();
```
@lutovich wouldn't that throw and exception when called from within the `EventLoop` as it may possible deadlock ?

lutovich replied:
Previous code executed `channel.close().awaitUninterruptibly()` unconditionally. This can be a problem, I think. Should be enough to not call `awaitUninterruptibly()` when in event loop that is responsible for `Future` returned by `closeIdleChannels()`, right? Will fix this

ejona86 replied:
A pet peeve of mine is when code behaves differently when running on the event loop...

In this code, implicitly no longer waiting for completion when on the event loop is broken, as the caller could have been expecting the blocking behavior (as that's what started this series of bug fixes!). When having one Channel interact with another Channel, it is random whether the two Channels share the same event loop. Code should either always block or never block, always schedule a task or never schedule a task, etc. I know that large portions of the Netty codebase don't agree with me on that :) (although that may change in Netty 5).

There's also a secondary fallacy involved here when checking inEventLoop(), since it uses await/syncUninterruptibly. Just because you aren't in _the_ event loop doesn't mean you aren't in _a_ event loop. It seems that issue was added in #7927 and it is possible to deadlock (if you are worried about being called from an event loop. Which is seems like you are concerned about that case since you checked inEventLoop to begin with).

I agree we're between a rock and a hard place, but I wonder if this is going about it the wrong way.

To me either 1) the method should always awaitUninterruptibly, and we just say you shouldn't call it from an event loop or 2) we make the method async again and solve the original bug another way. For (2) we could maybe create a new method that lets you wait until the cleanup is done by returning a future, or awaiting the pool's termination.

I may not understand these classes well enough, but it appears the code in this PR is even worse than the more general problem, because the `inEventLoop()` check is bonkers. It _looks_ like a normal check, but in actuality it doesn't protect anything. The `executor` was retrieved from `nextEventLoop()` which is gotten from `group.next()`. So it is a _random_ loop. Combined with the fact that we don't _directly_ schedule anything on that loop, it is sort of a senseless. Note that we do _indirectly_ schedule something on every event loop in the group: the listener. So if this was ever run from the event loop group it would deadlock (n-1)/n of the time (n is number of loops in group) and if run outside _this_ event loop group it would block still and thus could deadlock, but whether it actually deadlocks would depend on the full sum of behavior in the system.

lutovich replied:
@ejona86 I agree, changing method behavior depending on the caller thread is far from ideal. IMHO option (2) would be the best way to solve this and related problems. Not sure if it is fine wrt to backwards compatibility to restore the async behavior of `ChannelPool#close()` and add another interface method like `ChannelPool#closeFuture()`.

@normanmaurer what do you think?

lutovich commented on transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java:
```diff
@@ -389,13 +392,43 @@ protected boolean offerChannel(Channel channel) {
+                    channelsToClose.remove(channel);
+                    if (channelsToClose.isEmpty()) {
+                        // last channel was closed - can complete the result promise
+                        result.setSuccess(null);
```
Yeah, that's a problem. Will fix

lutovich commented:
@normanmaurer, @ejona86 review comments should now be addressed. I added separate commits for easier reviewing and can squash them when needed. Thanks!

normanmaurer commented:
@netty-bot test this please

normanmaurer commented on transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java:
```diff
@@ -389,13 +393,47 @@ protected boolean offerChannel(Channel channel) {
+        Promise<Void> result = executor.newPromise();
+        closeIdleChannels(result);
+        if (!executor.inEventLoop()) {
+            result.awaitUninterruptibly();
```
`sync...()` ?

normanmaurer commented on transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java:
```diff
@@ -389,13 +393,47 @@ protected boolean offerChannel(Channel channel) {
+        }
+        if (channelsToClose.isEmpty()) {
+            // no idle channels in the pool - nothing to close
+            result.setSuccess(null);
```
`trySuccess(...)` ?

ejona86 replied:
@normanmaurer, why was it suggested to swap to trySuccess here?

normanmaurer replied:
@ejona86 couldn't this race ?

ejona86 replied:
Doesn't look like it. It is a private method that is only used in one spot. The passed in promise is not used before this method and was not leaked to any other objects.

lutovich commented:
@normanmaurer comments should now be fixed

normanmaurer commented:
@netty-bot test this please...

normanmaurer commented:
@ejona86 mind reviewing again as well ?

carl-mastrangelo commented on transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java:
```diff
@@ -389,13 +393,47 @@ protected boolean offerChannel(Channel channel) {
+        Promise<Void> result = executor.newPromise();
+        closeIdleChannels(result);
+        if (!executor.inEventLoop()) {
+            result.syncUninterruptibly();
```
Sync uninterruptibly is not a good idea.  When the program is trying to shutdown (which is presumably when this method would be called), getting an Interruption means give up.  This should either declare it throws IE or wrap it in a runtime exception.

lutovich replied:
This method could exit before all the channels are closed if it is responsive to interrupts. Event loop threads will still be running even though the current thread is free to go. I'm not sure if it's okay or not. Re-interrupt + runtime exception would be the best option wrt to backward compatibility - in my opinion.

carl-mastrangelo replied:
SGTM

lutovich commented on transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java:
```diff
@@ -389,13 +393,47 @@ protected boolean offerChannel(Channel channel) {
+            channel.close().addListener(new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) {
+                    channelsToClose.remove(channel);
```
Could you please elaborate on why is this a bad idea? Set of channels is thread-safe and channels are removed from it by event loop threads that execute close future callbacks. It could happen that channel is removed by the current thread but this should not cause any problems.

I'm not sure how CountDownLatch can be used here. It only allows code to block waiting for the count to be zero. Blocking is not desirable here, I believe.

It could be an `AtomicInteger` for counting closed channels. However, I feel like the current code is a bit more explicit about what's going on. Set of channels to be closed is "snapshotted" and every closed channel is removed from the set.

lutovich commented on transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java:
```diff
@@ -389,13 +393,47 @@ protected boolean offerChannel(Channel channel) {
+                    if (channelsToClose.isEmpty()) {
+                        // last channel was closed - can complete the result promise
+                        // it might already be completed by a concurrently executing listener so use trySuccess
+                        result.trySuccess(null);
```
IMHO propagating idle channel closing errors from `SimpleChannelPool.close()` is not very caller-friendly. How would caller handle such error? Would it be useful to propagate an exception with 99 suppressed exceptions if pool had 100 idle channels and all of them failed to close?

Apache commons-pool swallows resource close errors when closed. Javadoc is [here](https://github.com/apache/commons-pool/blob/POOL_2_5_0/src/main/java/org/apache/commons/pool2/ObjectPool.java#L174) and impl is [here](https://github.com/apache/commons-pool/blob/POOL_2_5_0/src/main/java/org/apache/commons/pool2/impl/GenericObjectPool.java#L655-L667). HikariCP also just loggs such erros [here](https://github.com/brettwooldridge/HikariCP/blob/HikariCP-3.2.0/src/main/java/com/zaxxer/hikari/pool/HikariPool.java#L430-L447).

Callers that care about channel close errors can always subscribe to close future of every created channel and gather the errors. Logging errors in the pool could be nice to have.

carl-mastrangelo replied:
Since the Pool creates the channels, it's responsible for cleaning them up.  It would not be a correct abstraction boundary for callers to watch the close future for a Channel they didn't create.

lutovich commented:
@carl-mastrangelo thanks for reviewing the PR!

I'll wait a bit before addressing the comments because it is a bit unclear what is the best way forward, given concerns raised in https://github.com/netty/netty/pull/8408#discussion_r229727511.

lutovich commented:
Hey @normanmaurer, could you please review the discussion on this PR from https://github.com/netty/netty/pull/8408#discussion_r229727511 and assess what's the best way forward with this change? Thanks in advance!

lutovich commented:
Closing because a better fix was merged https://github.com/netty/netty/pull/9226.
